README
A little utility to make load time faster, especially by making

require 'xxx'

take much less time.  As in much less.  Well, on windows at least it's faster. 
It speeds up how long it takes things to load in windows, like rails apps.

Well, mostly on windows--on linux it's a speedup of only 0.41 to 0.45s, or so. [1]

If you've ever wondered why ruby feels slow on doze...sometimes it's just the startup time.  This really helps.

Benchmarks:

loading a (blank) rspec file:

  1.9.1   
    without 3.20s
    with 0.34s (10x improvement)

  1.8.6
    without 3.6s
    with 1.25s
    
    
rails app, running a unit test

 1.8.6
   without:
      23s
   with:
      14s

rails app, running $ script/console "puts 333"

 1.9.1
   without:
      20s
    with:
      10s
 
 1.8.6 
   without:
       9s
   with:
       6s
       
running "rake -T" somewhere

 1.9.1
   without: 3.75s
   with: 1.5s       

 1.8.6
   without: 1.37s
   with:    1.25s
   
Note: in reality what we should do is fix core so that it doesn't have such awful load time in windows.  There may be some inefficiency in there.  For now, this is a work-around that helps.

Also it helps with SSD's even, on windows, as it avoids some cpu used at require time, at least for 1.9.x

NB that installing newer versions of rubygems also seems to speedup start time (see below for how you can use them both, however, which is the best way).  
In fact, 1.8.7 with newer rubygems and only very few requires seems pretty fast, even without faster_require.
Once you have any number of requires, though, you'll want faster_require.
Rails needs it, and 1.9 still needs it, though, even with a newer rubygems, 
so you'll get some benefit from this library always--but don't take my word for it,
profile it and find out for yourself.

== How to use ==

The naive we is to use it by installing the gem, then adding

require 'rubygems'
require 'faster_require'

to the top of some (your initial) script.

However this doesn't speedup the loading of rubygems itself (XXX is much speedup?), so you can do this to get best performance:

G:>gem which faster_require
C:/installs/Ruby187/lib/ruby/gems/1.8/gems/faster_require-0.7.4/lib/faster_require.rb

now before doing require rubygems in your script, do this:

require 'C:/installs/Ruby187/lib/ruby/gems/1.8/gems/faster_require-0.7.4/lib/faster_require.rb'

Then rubygems' load will be cached too. Which is faster.  Or see "installing globally" below.

== How to use in Rails ==

One way is to install the gem, then add a

require 'rubygems'
require 'faster_require'

in your config/environment.rb, or (the better way is as follows):

Unpack it somewhere, like lib

$ cd my_rails_app/lib
$ gem unpack faster_require

Now add this line to your config/environment.rb:

require File.dirname(__FILE__) + "/../lib/faster_require-0.7.0/lib/faster_require" # faster speeds all around...make sure to update it to whatever version number you fetched though.

add that *before* this other (pre-existing) line:

require File.join(File.dirname(__FILE__), 'boot')

Now faster_require will speedup loading rubygems and everything railsy. Happiness.  (NB that setting it this will also run in production mode code, so be careful here, though it does work for me fine for production).

Ping me if it's still too slow.

== Clearing the cache ==

If you use bundler to change bundled gems, you'll want to run the command $ faster_require --clear-cache
so that it will pick up any new load paths.  Also if you moves files around to new directories, you may want to do the same.
As you install any new gems, it should clear the paths automatically for you.

== How to use generally/globally ==

You can install it to be used "always" (well, for anything that loads rubygems, at least, which is most things, via something like the following):

$ gem which rubygems
d:/Ruby192/lib/ruby/site_ruby/1.9.1/rubygems.rb

$ gem which faster_require
d:/Ruby192/lib/ruby/gems/1.9.1/gems/faster_require-0.6.0/lib/faster_require.rb

Now edit the rubygems.rb file, and add a require line to the top of it of the faster_require file, like this:
require 'd:/Ruby192/lib/ruby/gems/1.9.1/gems/faster_require-0.6.0/lib/faster_require.rb'

at the top of rubygems.rb file

update the path to be your own, obviously. You'll also have to change that added line if
you ever install a newer version of faster_require gem, or if you update your version of rubygems, 
as the rubygems.rb file will be wiped clean at that point.

This will cause everything to load faster.

== How to ignore PWD for faster_require ==

Faster_require also takes into consideration your Dir.pwd when you run it, for cacheing sake.
This means that if you run a ruby gem install script (like redcar's bin/redcar, for instance) and run it from different directories, it will always
be slow the first time you run it in each directory.
To make it be fast and basically disregard PWD, you can add global setting $faster_require_ignore_pwd_for_cache = true
and set it before requiring faster_require itself.

So now rubygems.rb at the top would look like
$faster_require_ignore_pwd_for_cache = true
require 'd:/Ruby192/lib/ruby/gems/1.9.1/gems/faster_require-0.6.0/lib/faster_require.rb'

And now your gem scripts will run fast regardless of where you run them from.

== See also ==

Spork can help speedup rails tests.  It "should" work well when used in conjunction with faster_require, as well.

Also "The Code Shop" releases versions of ruby that have an optimized require method, which might make this library obsolete.
http://thecodeshop.github.com
https://groups.google.com/group/thecodeshop
http://itreallymatters.net/post/12897174267/speedup-ruby-1-9-3-on-windows#.T89_XtVYut0

Enjoy.


ChangeLog
0.9.4: work with some UTF-8 encoded .rb files...which is the default now so hopefully ok.
  1.9.x compat. only [ping me if you want this changed]
  
0.9.3:
  possibly an execsjs fix?
  
0.9.1: work with rails 3.0.7 through a kludgey way...

0.9.0: refactors, bug fix for poor file, can autoload disregarding dir now 

0.8.1: work with 1.8.6 again

0.8.0: work with 1.9.x again

0.7.5: unique hash now, for the cache file [hopefuly]

0.7.4: allow double loading, for rails apps and gems to load it.

0.7.3: work with 1.9 require_relative

0.7.1: Accomodate for non HOME environment variable [what the..]

0.7.0: now it works with autoload finally [and thus hopefully rails]


spec/files/b.rb
if defined?($b) && $b > 0   
  raise 'cannot require b twice'
end
$b = 1

spec/files/c.rb
if defined?($c)
  raise 'cannot require twice'
else
  $c = true
end

spec/files/d.rb
require 'faster_require'

lib/fast_require.rb
module FastRequire
  $FAST_REQUIRE_DEBUG ||= false
  
  def self.setup
    @@dir = File.expand_path('~/.ruby_fast_require_cache')

    Dir.mkdir @@dir unless File.directory?(@@dir)
    @@loc = @@dir + '/' + RUBY_VERSION + '-' + RUBY_PLATFORM + '-' + 
       sanitize(File.expand_path($0).gsub(/[\/:]/, '_')) + 
       sanitize(Dir.pwd)
  end

  def self.sanitize filename
    filename.gsub(/[\/:]/, '_')
  end

  FastRequire.setup
  
  def self.load filename
    @@require_locs = Marshal.restore( File.open(filename, 'rb') {|f| f.read})
  end

  if File.exist?(@@loc)
    FastRequire.load @@loc
  else
    @@require_locs = {}
  end
  
  @@already_loaded = {}

  # try to see where this file was loaded from, from $:
  # partial_name might be abc.rb, or might be abc
  # partial_name might be a full path, too
  def self.guess_discover partial_name, add_dot_rb = false
    
    # test for full path first
    # unfortunately it has to be a full separate test
    # for windoze sake, as drive letter could be different than slapping a '/' on the dir to test list...
    tests = [partial_name]
    
    if add_dot_rb
      tests << partial_name + '.rb'
      tests << partial_name + '.so'
    end
    
    tests.each{|b|
      # assume that .rb.rb is...valid...
      if File.file?(b) && ((b[-3..-1] == '.rb') || (b[-3..-1] == '.so'))
        return File.expand_path(b)
      end
    }
      
    for dir in $:
      if File.file?(b = (dir + '/' + partial_name))
        # make sure we require a file that has the right suffix...
        if (b[-3..-1] == '.rb')  || (b[-3..-1] == '.so')
          return File.expand_path(b)
        end

      end
    end

    if add_dot_rb && (partial_name[-3..-1] != '.rb') && (partial_name[-3..-1] != '.so')
      guess_discover(partial_name + '.rb') || guess_discover(partial_name + '.so')
    else
      nil
    end
  end

  $LOADED_FEATURES.each{|already_loaded|
    # in 1.8 they might be partial paths
    # in 1.9, they might be non collapsed paths
    # so we have to sanitize them here...
    # XXXX File.exist? is a bit too loose, here...
    if File.exist?(already_loaded)
      key = File.expand_path(already_loaded)
    else
      key = FastRequire.guess_discover(already_loaded) || already_loaded
    end
    @@already_loaded[key] = true
  }
  
  @@already_loaded[File.expand_path(__FILE__)] = true # this file itself isn't in loaded features, yet, but very soon will be..
  # special case--I hope...
  
  # disallow re-requiring $0
  @@require_locs[$0] = File.expand_path($0) # so when we run into it on a require, we will skip it...
  @@already_loaded[File.expand_path($0)] = true
  
  
  # XXXX within a very long depth to require fast_require, 
  # require 'a' => 'b' => 'c' => 'd' & fast_require
  #             then
  #             => 'b.rb' 
  # it works always
  
  def self.already_loaded
    @@already_loaded
  end
  
  def self.require_locs
    @@require_locs
  end

  def self.dir
    @@dir
  end

  at_exit {
    FastRequire.default_save
  }
  
  def self.default_save
    self.save @@loc
  end

  def self.save to_file
    File.open(to_file, 'wb'){|f| f.write Marshal.dump(@@require_locs)}
  end

  def self.clear_all!
    require 'fileutils'
    FileUtils.rm_rf @@dir if File.exist? @@dir
    @@require_locs.clear
    setup
  end

  def require_cached lib
    lib = lib.to_s unless lib.is_a?(String) # might not be zactly 1.9 compat...
    if known_loc = @@require_locs[lib]
      return false if @@already_loaded[known_loc]
      @@already_loaded[known_loc] = true
      if known_loc =~ /.so$/
        puts 'doing original_non_cached_require on .so full path ' + known_loc if $FAST_REQUIRE_DEBUG
        original_non_cached_require known_loc # not much we can do there...too bad...
      else
        puts 'doing eval on ' + lib + '=>' + known_loc if $FAST_REQUIRE_DEBUG
        $LOADED_FEATURES << known_loc # *must*
        return eval(File.open(known_loc, 'rb') {|f| f.read}, TOPLEVEL_BINDING, known_loc) || true # note the b here--this means it's reading .rb files as binary, which *typically* works--if it breaks re-save the offending file in binary mode, or file an issue on the tracker...
      end
    else
      # we don't know the location--let Ruby's original require do the heavy lifting for us here
      old = $LOADED_FEATURES.dup
      if(original_non_cached_require lib)
        # debugger might land here the first time you run a script and it doesn't have a require
        # cached yet...
        new = $LOADED_FEATURES - old
        found = new.last

        # incredibly, in 1.8.6, this doesn't always get set to a full path
        if RUBY_VERSION < '1.9'
          if !File.file?(found)
            # discover the full path.
            dir = $:.find{|path| File.file?(path + '/' + found)}
            found = dir + '/' + found
          end
          found = File.expand_path(found);
        end
        puts 'found new loc:' + lib + '=>' + found if $FAST_REQUIRE_DEBUG
        @@require_locs[lib] = found
        @@already_loaded[found] = true
        return true
      else
        puts 'already loaded, apparently' + lib if $FAST_REQUIRE_DEBUG
        # this probably was something like
        # the first pass was require 'regdeferred'
        # now it's a different require 'regdeferred.rb'
        # which fails (or vice versa)
        # so figure out why
        # calc location, expand, map back
        where_found = FastRequire.guess_discover(lib, true)
        if where_found
          puts 'inferred ghost loc:' + lib + '=>' + where_found if $FAST_REQUIRE_DEBUG
          @@require_locs[lib] = where_found
          # unfortunately if it's our first pass
          # and we are in the middle of a "real" require
          # that is circular
          # then $LOADED_FEATURES or (AFAIK) nothing will have been set
          # for us to be able to assert that
          # so...I think we'll end up
          # just fudging for a bit
          #	raise 'not found' unless @@already_loaded[where_found] # should have already been set...I think...
        else
          if $FAST_REQUIRE_DEBUG
            # happens for enumerator XXXX
            puts 'unable to infer' + lib + ' in ' if $FAST_REQUIRE_DEBUG
            @@already_loaded[found] = true # hacky
          end
        end
        return false # XXXX test all these return values
      end
    end
  end

  def self.resetup!
    eval "module ::Kernel; alias :require :require_cached; end"
  end
end

module Kernel

  if(defined?(@already_using_fast_require))
    raise 'twice not allowed...'
    # *shouldn't* get here...unless I'm wrong...
  else
    @already_using_fast_require = true
    include FastRequire
    # overwrite old require...
    alias :original_non_cached_require :require
    FastRequire.resetup!
  end

end


lib/faster_require.rb
# check for and avoid a double load...because we're afraid to load twice, since we override require et al

if(defined?($already_using_faster_require))
  p 'warning: faster_require double load--expected?' if $FAST_REQUIRE_DEBUG
  local_version = File.read(File.dirname(__FILE__) + "/../VERSION")
  raise "mismatched faster_require versions! #{local_version} != #{FastRequire::VERSION}" unless local_version == FastRequire::VERSION
else
  
$already_using_faster_require = true


require 'rbconfig' # maybe could cache this one's loc, too? probably not...

module FastRequire
  
  $FAST_REQUIRE_DEBUG ||= $DEBUG # can set this via $DEBUG, or on its own previously
  
  VERSION = File.read(File.dirname(__FILE__) + "/../VERSION")

  def self.sanitize filename
    filename.gsub(/[\/:]/, '_')
  end
  
  if RUBY_VERSION >= '1.9.0'
    # appears 1.9.x has inconsistent string hashes...so roll our own...
    def self.string_array_cruddy_hash strings
      # we only call this method once, so overflowing to a bignum is ok
      hash = 1;
      for string in strings
        hash = hash * 31
        string.each_byte{|b|
          hash += b
        }
      end
      hash # probably a Bignum (sigh)
    end
    
  else
    
    def self.string_array_cruddy_hash strings
      strings.hash
    end
    
  end

  def self.setup
    begin
     @@dir = File.expand_path('~/.ruby_faster_require_cache')
    rescue ArgumentError => e # couldn't find HOME environment or the like
     whoami = `whoami`.strip
     if File.directory?(home = "/home/#{whoami}")
      @@dir = home + '/.ruby_faster_require_cache'
     else
       raise e.to_s + " and couldnt infer it from whoami"
     end
    end

    unless File.directory?(@@dir)
      Dir.mkdir @@dir 
      raise 'unable to create user dir for faster_require ' + @@dir unless File.directory?(@@dir)
    end
    
    config = RbConfig::CONFIG
    
    # try to be a unique, but not too long, filename, for restrictions on filename length in doze
    ruby_bin_name = config['bindir'] + config['ruby_install_name'] # needed if you have two rubies, same box, same ruby description [version, patch number]
    parts = [File.basename($0), RUBY_PATCHLEVEL.to_s, RUBY_PLATFORM, RUBY_VERSION, RUBY_VERSION, File.expand_path(File.dirname($0)), ruby_bin_name]
    unless defined?($faster_require_ignore_pwd_for_cache)
      # add in Dir.pwd
      parts << File.basename(Dir.pwd)
      parts << Dir.pwd
    else
      p 'ignoring dirpwd for cached file location' if $FAST_REQUIRE_DEBUG
    end
    
    sanitized_parts = parts.map{|part| sanitize(part)}

    full_parts_hash = string_array_cruddy_hash(parts).to_s
    
    loc_name = (sanitized_parts.map{|part| part[0..5] + (part[-5..-1] || '')}).join('-') + '-' + full_parts_hash + '.marsh'
    
    @@loc = @@dir + '/' + loc_name
    
    if File.exist?(@@loc)
      FastRequire.load @@loc
    else
      @@require_locs = {}
    end
      
    @@already_loaded = {}
  
    $LOADED_FEATURES.each{|already_loaded|
      # in 1.8 they might be partial paths
      # in 1.9, they might be non collapsed paths
      # so we have to sanitize them here...
      # XXXX File.exist? is a bit too loose, here...
      if File.exist?(already_loaded)
        key = File.expand_path(already_loaded)
      else
        key = FastRequire.guess_discover(already_loaded) || already_loaded
      end
      @@already_loaded[key] = true
    }
  
    @@already_loaded[File.expand_path(__FILE__)] = true # this file itself isn't in loaded features, yet, but very soon will be..
    # a special case--I hope...
  
    # also disallow re-loading $0
    @@require_locs[$0] = File.expand_path($0) # so when we run into $0 on a freak require, we will skip it...
    @@already_loaded[File.expand_path($0)] = true
    
  end
  
  def self.load filename
    cached_marshal_data = File.open(filename, 'rb') {|f| f.read}
    begin
      @@require_locs = Marshal.restore( cached_marshal_data )
    rescue ArgumentError
      @@require_locs= {}
    end
  end


  # try to see where this file was loaded from, from $:
  # partial_name might be abc.rb, or might be abc
  # partial_name might be a full path, too
  def self.guess_discover partial_name, add_dot_rb = false

    # test for full path first
    # unfortunately it has to be a full separate test
    # for windoze sake, as drive letter could be different than slapping a '/' on the dir to test list...
    tests = [partial_name]

    if add_dot_rb
      tests << partial_name + '.rb'
      tests << partial_name + '.' + RbConfig::CONFIG['DLEXT']
    end

    tests.each{|b|
      # assume that .rb.rb is...valid...?
      if File.file?(b) && ((b[-3..-1] == '.rb') || (b[-3..-1] == '.' + RbConfig::CONFIG['DLEXT']))
        return File.expand_path(b)
      end
    }

    for dir in $:
      if File.file?(b = (dir + '/' + partial_name))
        # make sure we require a file that has the right suffix...
        if (b[-3..-1] == '.rb')  || (b[-3..-1] == '.' + RbConfig::CONFIG['DLEXT'])
          return File.expand_path(b)
        end

      end
    end

    if add_dot_rb && (partial_name[-3..-1] != '.rb') && (partial_name[-3..-1] != '.' + RbConfig::CONFIG['DLEXT'])
      guess_discover(partial_name + '.rb') || guess_discover(partial_name + '.')
    else
      nil
    end
  end
  
  FastRequire.setup
  
  def self.already_loaded
    @@already_loaded
  end

  def self.require_locs
    @@require_locs
  end

  def self.dir
    @@dir
  end
  
  def self.loc
    @@loc
  end

  at_exit {
    FastRequire.default_save
  }

  def self.default_save
    self.save @@loc
  end

  def self.save to_file
    File.open(to_file, 'wb'){|f| f.write Marshal.dump(@@require_locs)}
  end

  # for testing use only, basically
  def self.clear_all!
    require 'fileutils'
    success = false
    if File.exist? @@dir
      FileUtils.rm_rf @@dir 
      success = true
    end
    @@require_locs.clear
    setup
    success
  end
  
  private
  def last_caller
   caller[-2]
  end
  
  IN_PROCESS = []
  ALL_IN_PROCESS = []
  @@count = 0
  
  public
  
  def require_cached lib
    lib = lib.to_s # might not be zactly 1.9 compat... to_path ??
    ALL_IN_PROCESS << [lib, @@count += 1]
    begin
      p 'doing require ' + lib + ' from ' + caller[-1] if $FAST_REQUIRE_DEBUG
      if known_loc = @@require_locs[lib]
        if @@already_loaded[known_loc]
          p 'already loaded ' + known_loc + ' ' + lib if $FAST_REQUIRE_DEBUG
          return false 
        end
        @@already_loaded[known_loc] = true
        if known_loc =~ /\.#{RbConfig::CONFIG['DLEXT']}$/
          puts 'doing original_non_cached_require on .so full path ' + known_loc if $FAST_REQUIRE_DEBUG
          original_non_cached_require known_loc # not much we can do there...too bad...well at least we pass it a full path though :P
        else
          unless $LOADED_FEATURES.include? known_loc
            if known_loc =~ /rubygems.rb$/
              puts 'requiring rubygems ' + lib if $FAST_REQUIRE_DEBUG
              original_non_cached_require(lib) # revert to normal require so rubygems doesn't freak out when it finds itself already in $LOADED_FEATURES with rubygems > 1.6 :P
            else
              IN_PROCESS << known_loc
              begin
                if $FAST_REQUIRE_DEBUG
                  puts 'doing cached loc eval on ' + lib + '=>' + known_loc + " with stack:" + IN_PROCESS.join(' ')
                end
                $LOADED_FEATURES << known_loc
                # fakely add the load path, too, so that autoload for the same file/path in gems will work <sigh> [rspec2]
                no_suffix_full_path = known_loc.gsub(/\.[^.]+$/, '')
                no_suffix_lib = lib.gsub(/\.[^.]+$/, '')
                libs_path = no_suffix_full_path.gsub(no_suffix_lib, '')
                libs_path = File.expand_path(libs_path) # strip off trailing '/'
                
                $: << libs_path unless $:.index(libs_path) # add in this ones real require path, so that neighboring autoloads will work
                known_locs_dir = File.dirname(known_loc)
                $: << known_locs_dir unless $:.index(known_locs_dir) # attempt to avoid the regin loading bug...yipes.
                
                # try some more autoload conivings...so that it won't attempt to autoload if it runs into it later...
                relative_full_path = known_loc.sub(libs_path, '')[1..-1]
                $LOADED_FEATURES << relative_full_path unless $LOADED_FEATURES.index(relative_full_path) # add in with .rb, too, for autoload 
                  
                # load(known_loc, false) # too slow
                
                # use eval: if this fails to load breaks re-save the offending file in binary mode, or file an issue on the faster_require tracker...
                contents = File.open(known_loc, 'rb:utf-8') {|f| f.read} # read only costs 0.34/10.0 s...
                if contents =~ /require_relative/ # =~ is faster apparently faster than .include?
                  load(known_loc, false) # load is slow, but overcomes a ruby core bug: http://redmine.ruby-lang.org/issues/4487
                else
                  eval(contents, TOPLEVEL_BINDING, known_loc)
                end
              ensure
                raise 'unexpected' unless IN_PROCESS.pop == known_loc
              end
              return true
            end
          else
            puts 'ignoring already loaded [circular require?] ' + known_loc + ' ' + lib if $FAST_REQUIRE_DEBUG
          end
        end
      else
        # we don't know the location--let Ruby's original require do the heavy lifting for us here
        old = $LOADED_FEATURES.dup
        p 'doing old non-known location require ' + lib if $FAST_REQUIRE_DEBUG
        if(original_non_cached_require(lib))
          # debugger might land here the first time you run a script and it doesn't have a require
          # cached yet...
          new = $LOADED_FEATURES - old
          found = new.last
  
          # incredibly, in 1.8.x, this doesn't always get set to a full path.
          if RUBY_VERSION < '1.9'
            if !File.file?(found)
              # discover the full path.
              dir = $:.find{|path| File.file?(path + '/' + found)}
              return true unless dir # give up, case jruby socket.jar "mysterious"
              found = dir + '/' + found
            end
            found = File.expand_path(found);
          end
          puts 'found new loc:' + lib + '=>' + found if $FAST_REQUIRE_DEBUG
          @@require_locs[lib] = found
          @@already_loaded[found] = true
          return true
        else
        
          # this is expected if it's for libraries required before faster_require was [like rbconfig]
          # raise 'actually expected' + lib if RUBY_VERSION >= '1.9.0'
          puts 'already loaded, apparently [require returned false], trying to discover how it was redundant... ' + lib if $FAST_REQUIRE_DEBUG
          # this probably was something like
          # the first pass was require 'regdeferred'
          # now it's a different require 'regdeferred.rb'
          # which fails (or vice versa)
          # so figure out why
          # calc location, expand, map back
          where_found = FastRequire.guess_discover(lib, true)
          if where_found
            puts 'inferred lib loc:' + lib + '=>' + where_found if $FAST_REQUIRE_DEBUG
            @@require_locs[lib] = where_found
            # unfortunately if it's our first pass
            # and we are in the middle of a "real" require
            # that is circular
            # then $LOADED_FEATURES or (AFAIK) nothing will have been set
            # for us to be able to assert that
            # so...I think we'll end up
            # just fudging for a bit
            #	raise 'not found' unless @@already_loaded[where_found] # should have already been set...I think...
          else
            if $FAST_REQUIRE_DEBUG
              # happens for enumerator XXXX
              puts 'unable to infer ' + lib + ' location' if $FAST_REQUIRE_DEBUG
              @@already_loaded[found] = true # so hacky...
            end
          end
          return false # XXXX test all these return values
        end
      end
    ensure
      raise 'huh' unless ALL_IN_PROCESS.pop[0] == lib
    end
  end

end

module Kernel

  # overwrite old require...
  include FastRequire
  if defined?(gem_original_require)
    class << self
      alias :original_remove_method :remove_method
      
      def remove_method method # I think this actually might be needed <sigh>
        if method.to_s == 'require'
          #p 'not removing old require, since that\'s ours now'
        else
          original_remove_method method
        end
      end
    
    end
    
    # similarly overwrite this one...I guess...1.9.x...rubygems uses this as its default...I think...
    alias :original_non_cached_require :gem_original_require
    alias :gem_original_require :require_cached
  else
    alias :original_non_cached_require :require
    alias :require :require_cached
  end
end

end
